Fails to compile on ESPHome 2022.12.3  (fix included)

I only found this because I noticed the copy of HomeAssistant on my Home Assistant on my local instance was stuck at version 2022.3.1. The version in the Add-ons Store was 2022.12.x but had moved repository -- I had to uninstall it, then reinstall it from the store to make it update.

After updating, the Flipdot driver failed to build with this error:

```
Compiling /data/flipdot-small/.pioenvs/flipdot-small/lib64d/WiFi/WiFiClient.cpp.o
src/main.cpp: In function 'void setup()':
src/main.cpp:253:51: error: invalid new-expression of abstract class type 'esphome::hanover_flipdot::hanover_flipdot'
   flipdot1 = new hanover_flipdot::hanover_flipdot();
                                                   ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/hanover_flipdot/hanover_flipdot.h:10:7: note:   because the following virtual functions are pure within 'esphome::hanover_flipdot::hanover_flipdot':
 class hanover_flipdot : public PollingComponent, public display::DisplayBuffer, public uart::UARTDevice  {
       ^~~~~~~~~~~~~~~
In file included from src/esphome.h:16,
                 from src/main.cpp:3:
src/esphome/components/display/display_buffer.h:375:23: note: 	'virtual esphome::display::DisplayType esphome::display::DisplayBuffer::get_display_type()'
   virtual DisplayType get_display_type() = 0;
                       ^~~~~~~~~~~~~~~~
Compiling /data/flipdot-small/.pioenvs/flipdot-small/lib64d/WiFi/WiFiGeneric.cpp.o
*** [/data/flipdot-small/.pioenvs/flipdot-small/src/main.cpp.o] Error 1
```

This seems to be due to an API change in the ESPHome display component. Adding this line to the `public` section of the `hanover_flipdot` class (above the `protected:` block) in `hanover_flipdot.h` fixes it:

```
  display::DisplayType get_display_type() override { return  display::DisplayType::DISPLAY_TYPE_BINARY; }
```

--- a/hanover_flipdot/display.py
+++ b/hanover_flipdot/display.py
@@ -9,6 +9,7 @@ from esphome.const import (
     CONF_PAGES,
     CONF_ADDRESS,
 )
+from esphome.const import __version__ as ESPHOME_VERSION
 
 CODEOWNERS = ["gattrill"]
 DEPENDENCIES = ["uart"]
@@ -47,7 +48,8 @@ async def to_code(config):
     cg.add(var.set_address(config[CONF_ADDRESS]))
     cg.add(var.set_offset(config[CONF_OFFSET]))
 
-    await cg.register_component(var, config)
+    if cv.Version.parse(ESPHOME_VERSION) < cv.Version.parse("2023.12.0"):
+        await cg.register_component(var, config)
     await display.register_display(var, config)
 
     if CONF_OFFSET in config:

--- a/hanover_flipdot/hanover_flipdot.h
+++ b/hanover_flipdot/hanover_flipdot.h
@@ -27,6 +27,10 @@ class hanover_flipdot : public PollingComponent, public display::DisplayBuffer, 
   void setup() override {
     this->initialize();
   }
+ 
+  display::DisplayType get_display_type() override {
+   return display::DisplayType::DISPLAY_TYPE_BINARY;
+  }
 
  protected:
   void draw_absolute_pixel_internal(int x, int y, Color color) override;

--- a/hanover_flipdot/hanover_flipdot.h
+++ b/hanover_flipdot/hanover_flipdot.h
@@ -1,13 +1,18 @@
 #pragma once
 
 #include "esphome/core/component.h"
+#include "esphome/core/version.h"
 #include "esphome/components/uart/uart.h"
 #include "esphome/components/display/display_buffer.h"
 
 namespace esphome {
 namespace hanover_flipdot {
 
+#if ESPHOME_VERSION_CODE >= VERSION_CODE(2023, 12, 0)
+class hanover_flipdot : public display::DisplayBuffer, public uart::UARTDevice  {
+#else
 class hanover_flipdot : public PollingComponent, public display::DisplayBuffer, public uart::UARTDevice  {
+#endif  // VERSION_CODE(2023, 12, 0)
  public:
 
   float get_setup_priority() const override { return setup_priority::PROCESSOR; }

